lib/IRGen/StackSizeLimitTransform.cpp

```diff
@@ -0,0 +1,307 @@
+processAllocRefInst(StackSizeLimitState &state, AllocRefInst *ari,
+                    SILBasicBlock::iterator next) {
+  // Skip alloc_ref that are not marked as can alloc stack or ones that
+  // are marked as is objc.
+  if (!ari->canAllocOnStack() || ari->isObjC()) {
+    // We can not stack allocate objective-c allocations.
+    state.madeChange = true;
+    ari->setStackAllocatable(false);
+    return next;
+  }
+
+  SILType selfType = ari->getType();
+  auto &classTI = state.igm.getTypeInfo(selfType).as<irgen::ClassTypeInfo>();
+  auto &classLayout = classTI.getClassLayout(state.igm, selfType,
+                                             /*forBackwardDeployment=*/false);
+  if (!classLayout.isFixedLayout()) {
+    // We never stack allocate if we don't have a fixed layout.
+    state.madeChange = true;
+    ari->setStackAllocatable(false);
+    return next;
+  }
+
+  // Calculate the total size needed.
+  // The first part is the size of the class itself.
+  Alignment classAlignment = classLayout.getAlignment();
+  Size accumulatedSize = classLayout.getSize();
+
+  auto tailTypeArray = ari->getTailAllocatedTypes();
+  auto tailCountArray = ari->getTailAllocatedCounts();
+  for (auto p : llvm::zip(tailTypeArray, tailCountArray)) {
+    SILType eltType = std::get<0>(p);
+    Operand &operand = std::get<1>(p);
+    auto *count = dyn_cast<IntegerLiteralInst>(operand.get());
+    if (!count) {
+      state.madeChange = true;
+      ari->setStackAllocatable(false);
+      return next;
+    }
+
+    const TypeInfo &elemTI = state.igm.getTypeInfo(eltType);
+    if (!elemTI.isFixedSize()) {
+      state.madeChange = true;
+      ari->setStackAllocatable(false);
+      return next;
+    }
+
+    const FixedTypeInfo &elemFTI = elemTI.as<FixedTypeInfo>();
+    Alignment elemAlign = elemFTI.getFixedAlignment();
+
+    // This should not happen - just to be safe, mark the whole
+    // allocation as on the heap.
+    if (elemAlign > classAlignment) {
+      state.madeChange = true;
+      ari->setStackAllocatable(false);
+      return next;
+    }
+
+    accumulatedSize = accumulatedSize.roundUpToAlignment(elemAlign);
+    accumulatedSize +=
+        count->getValue().getZExtValue() * elemFTI.getFixedStride();
+  }
+
+  if (accumulatedSize > Size(state.stackAllocSize)) {
+    state.madeChange = true;
+    ari->setStackAllocatable(false);
+    return next;
+  }
+
+  // Otherwise, we can stack allocate.
+  state.stackAllocSize -= accumulatedSize.getValue();
+  return next;
+}
+
+//===----------------------------------------------------------------------===//
+//                            Top Level Entrypoint
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+class StackSizeLimitTransform : public SILFunctionTransform {
+  void run() override {
+    auto *f = getFunction();
+    auto *igm = getIRGenModule();
+    StackSizeLimitState state(igm);
+    bool madeChange = false;
```

Comment:
This pass should not run on async functions (except for lowering the builtin).
The "stack" for async functions is heap allocated anyway, so it doesn't harm if there are large objects on the async "stack".

Reply:
Makes sense. Will fix! And will have a test for it!